d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -166,29 +177,38 @@ String getXdsServerAuthority()
    * NOTE: Must be called from the executor.
    * @return {@code true} if the client is in backoff
    */
-  private boolean isInBackoff() {
+  private boolean isInBackoff()
+  {
     return _adsStream == null && _retryRpcStreamFuture != null && !_retryRpcStreamFuture.isDone();
   }
 
   /**
    * Handles ready callbacks from the RPC stream. Must be called from the executor.
    */
-  private void readyHandler() {
+  private void readyHandler()
+  {
     _log.debug("Received ready callback from the ADS stream");
-    if (_adsStream == null || isInBackoff()) {
+    if (_adsStream == null || isInBackoff())
+    {
       _log.warn("Unexpected state, ready called on null or backed off ADS stream!");
       return;
     }
     // confirm ready state to neglect spurious callbacks; we'll get another callback whenever it is ready again.
-    if (_adsStream.isReady()) {
+    if (_adsStream.isReady())
+    {
       // if the ready timeout future is non-null, a reconnect notification hasn't been sent yet.
-      if (_readyTimeoutFuture != null) {
+      if (_readyTimeoutFuture != null)
+      {
         // timeout task will be cancelled only if it hasn't already executed.
         boolean cancelledTimeout = _readyTimeoutFuture.cancel(false);
         _log.info("ADS stream ready, cancelled timeout task: {}", cancelledTimeout);
         _readyTimeoutFuture = null; // set it to null to avoid repeat notifications to subscribers.
+        _xdsClientJmx.incrementReconnectionCount();
         notifyStreamReconnect();
+        // note: no need to start a retry task explicitly since xds stream internally will keep on retrying to connect
+        // to one of the sub-channels (unless an error or complete callback is called).
```

Comment:
Maybe this should go here? https://github.com/linkedin/rest.li/blob/559f619ba0b68873ec35b8e612206b6a9af1a375/d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java#L150
Thanks for adding it btw.

Reply:
whoops, yeah, wrong place. Corrected.